File: CPP/Tree/BiTreeTraverse_Recursive.cpp
#include "../Basic_Structure.h"

#include <iostream>

// 二叉树的三种遍历顺序

void preOrder(BiTreeNode* T){
    if(T == nullptr){
        return;
    }
    std::cout << T->val << "-";
    preOrder(T->left);
    preOrder(T->right);
}

void inOrder(BiTreeNode* T){
    if(T == nullptr){
        return;
    }
    preOrder(T->left);
    std::cout << T->val << "-";
    preOrder(T->right);
}

void postOrder(BiTreeNode* T){
    if(T == nullptr){
        return;
    }
    preOrder(T->left);
    preOrder(T->right);
    std::cout << T->val << "-";
}

void f(BiTreeNode* T){
    if(T == nullptr){
        return;
    }
    // #1
    f(T);
    // #2
    f(T);
    // #3
    // 在函数调用栈中, 每个节点会出现3次, 3种不同的遍历顺序就对应了节点出现的第几次打印
}

File: Python/Tree/BiTreeTraverse_Recursive.py
import sys
import os
# 将项目根目录添加到 Python 路径中
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from Basic_Structure import BiTreeNode

# 先序遍历
def preOrder(T: BiTreeNode) -> list:
    if not T:
        return []
    return [T.value] + preOrder(T.left) + preOrder(T.right)

# 中序遍历
def inOrder(T: BiTreeNode) -> list:
    if not T:
        return []
    return inOrder(T.left) + [T.value] + inOrder(T.right)

# 后序遍历
def postOrder(T: BiTreeNode) -> list:
    if not T:
        return []
    return postOrder(T.left) + postOrder(T.right) + [T.value]